### ISSUE TITLE
train existing existing engine with additional parallel data

### ISSUE THREAD

**Issue by philippehenri** [ModernMT/MMT#393]
Hi

once a neural engine is created (./mmt create en fr examples/data -e eng1 --neural)

- Is it possible to train with additional data? (am not taking about importing into memory)
- will a dictionary improve translations (if added to training data)?

thanks
philippe

**Comment by davidecaroselli** [comment]
Hi @philippehenri 

the way ModernMT handles new data **after** the training of the engine is with `./mmt import`, in other words, adding data to the internal memory and use it for adaptation.

More in general there is no "easy" way to add new data directly to the neural engine after the training is completed. We solved this problem with ModernMT's Adaptivity feature. The only good advice I can give you is to have a "solid" and large baseline model.

Davide

**Comment by philippehenri** [comment]
Hi Davide

thanks for your reply.

and about the dictionary? if we add a multilingual dictionary/glossary (with entries as one or more words) to the set of training files, will that improve the translations?
i have seen in a thread that somebody was suggesting that chunks of text was improving results
thanks
philippe

**Comment by EtienneAb3d** [comment]
It may improve. But..

The problem with dictionaries are:
1) the words/expressions are out of context, certainly in a normalized form only, while the neural net is learning how to phrase, thus, need in-context flexed words, 
2) dictionaries are often providing a small amount of expressions, all enumerated only once, while the neural net needs a statistical effect to be efficient. 

The advantage of the chunking suggestion are:
1) the chunks are all pieces of sentences, containing local in-context flexed words
2) the chunk extraction provide with a very large amount of data, nearly as large, and as redundant, as your original learning sentences

**Comment by philippehenri** [comment]
Hi Etienne,

"...while the neural net needs a statistical effect to be efficient"

So if I understand well, it means that  glossaries with multi-words entries could still be useful (as they create some kind of context, ex: "juge de paix" in french) - vs one word entries (which do not)?

And if we want these important terms to have a higher statistical effect (as these terms normally appear once in the glossary), what will be the effect of duplicating the glossaries in the training set (to cheat with the statistics)...

Has someone experience doing this?

thanks
philippe

**Comment by EtienneAb3d** [comment]
To be more precise, here is the list of chunks I got on my side for "juge de paix", on usual available resources:
```
europarl.fr-en.chunks.fr:peut agir comme un juge de paix
europarl.fr-en.chunks.fr:appel au juge de paix
europarl.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:de vice-maire et de juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:( juge de paix ) .
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:d' un juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix Fitzmaurice
MultiUN.fr-en.chunks.fr:Troisième juge de paix de la branche civile
MultiUN.fr-en.chunks.fr:le juge de paix siégeant en tribunal de police
MultiUN.fr-en.chunks.fr:du juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:dans la compétence du juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:Le tribunal de simple police est composé du juge de paix
MultiUN.fr-en.chunks.fr:les plus couramment employées devant le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix assisté de quatre assesseurs choisis en
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:à moins que le juge de paix
MultiUN.fr-en.chunks.fr:président d' un tribunal de première instance ou juge de paix de procéder
MultiUN.fr-en.chunks.fr:du ressort du juge de paix
MultiUN.fr-en.chunks.fr:doit être confirmée par un juge de paix (
MultiUN.fr-en.chunks.fr:le juge de paix (
MultiUN.fr-en.chunks.fr:Le jour suivant , le juge de paix a prononcé
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:matière d' instruction du juge de paix
MultiUN.fr-en.chunks.fr:au juge de paix
MultiUN.fr-en.chunks.fr:l' issue d' une demande verbale faite au juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix ou
MultiUN.fr-en.chunks.fr:le juge de paix de Delmas
MultiUN.fr-en.chunks.fr:profit d' un juge de paix suppléant de Delmas
MultiUN.fr-en.chunks.fr:la place d' un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix de
MultiUN.fr-en.chunks.fr:et le juge de paix
MultiUN.fr-en.chunks.fr:poste de juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:comporte le juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:tribunal cantonal par l' intermédiaire du juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix dans les cas où
MultiUN.fr-en.chunks.fr:au juge de paix de Pétion-Ville aux fins de
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:poste de juge de paix communautaire
MultiUN.fr-en.chunks.fr:poste de juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix de
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:poste de juge de paix
MultiUN.fr-en.chunks.fr:par le juge de paix siégeant en
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:un total de 318 juge de paix
MultiUN.fr-en.chunks.fr:dirigé à Oslo par le juge de paix
MultiUN.fr-en.chunks.fr:le cas de l' ancien juge de paix de Santa Lucía de Utatlán
MultiUN.fr-en.chunks.fr:a pris d' assaut les locaux du tribunal et roué de coups le juge de paix Alvaro Martínez
MultiUN.fr-en.chunks.fr:déjà mentionnés du lynchage du juge de paix de Senahú
MultiUN.fr-en.chunks.fr:du juge de paix
MultiUN.fr-en.chunks.fr:Le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix peut ordonner
MultiUN.fr-en.chunks.fr:Le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:décision du 13 février 2004 du juge de paix de Bäretswil
MultiUN.fr-en.chunks.fr:décision du 13 février 2004 du juge de paix de Bäretswil
MultiUN.fr-en.chunks.fr:la charge de juge de paix
MultiUN.fr-en.chunks.fr:la notion de juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:une nouvelle évaluation des éléments de preuve examinés par le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:juge de paix du lieu de résidence
MultiUN.fr-en.chunks.fr:au juge de paix du lieu
MultiUN.fr-en.chunks.fr:une visite préalable du juge de paix
MultiUN.fr-en.chunks.fr:au juge de paix
MultiUN.fr-en.chunks.fr:Le juge de paix doit
MultiUN.fr-en.chunks.fr:décision du juge de paix
MultiUN.fr-en.chunks.fr:les jugements du juge de paix rendus en
MultiUN.fr-en.chunks.fr:la fonction de juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix dresse
MultiUN.fr-en.chunks.fr:Le juge de paix
MultiUN.fr-en.chunks.fr:Le juge de paix siège seul
MultiUN.fr-en.chunks.fr:secrétaire ou à un employé du juge de paix
MultiUN.fr-en.chunks.fr:à celui du juge de paix Miguel Geovanni Avila Vásquez
MultiUN.fr-en.chunks.fr:du juge de paix Alvaro Hugo Martínez , à Senahú
MultiUN.fr-en.chunks.fr:juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:juge de paix de prendre
MultiUN.fr-en.chunks.fr:une nomination de juge de paix
MultiUN.fr-en.chunks.fr:Le juge de paix
MultiUN.fr-en.chunks.fr:Le juge de paix
MultiUN.fr-en.chunks.fr:extension de la tâche du juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix
MultiUN.fr-en.chunks.fr:du juge de paix
MultiUN.fr-en.chunks.fr:celle d' un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:un juge de paix
MultiUN.fr-en.chunks.fr:trois du lynchage d' un juge de paix en 2001
MultiUN.fr-en.chunks.fr:ministre du culte ou juge de paix
MultiUN.fr-en.chunks.fr:, ou un juge de paix connaissant
MultiUN.fr-en.chunks.fr:juge de paix
MultiUN.fr-en.chunks.fr:compétences du juge de paix
MultiUN.fr-en.chunks.fr:au juge de paix de prononcer l' extinction de l' action pénale
MultiUN.fr-en.chunks.fr:Le juge de paix convoque la victime
MultiUN.fr-en.chunks.fr:vertu d' un mandat délivré par un juge de paix
MultiUN.fr-en.chunks.fr:le juge de paix où le ministère public
MultiUN.fr-en.chunks.fr:Le juge de paix dans les justices de paix
MultiUN.fr-en.chunks.fr:peut être également nommé aux fonctions de juge de paix
MultiUN.fr-en.chunks.fr:organisation judiciaire aux termes duquel les fonctions de juge de paix
Wikipedia.fr-en.chunks.fr:a été élu maire de Berkeley et juge de paix
Wikipedia.fr-en.chunks.fr:Il est nommé à un poste nouvellement créé de juge de paix
Wikipedia.fr-en.chunks.fr:le juge de paix Mordecai Lincoln
Wikipedia.fr-en.chunks.fr:un juge de paix
Wikipedia.fr-en.chunks.fr:le juge de paix de Lambesc
Wikipedia.fr-en.chunks.fr:, le juge de paix Pagnot
Wikipedia.fr-en.chunks.fr:le juge de paix exécuté
Wikipedia.fr-en.chunks.fr:juge de paix
Wikipedia.fr-en.chunks.fr:juge de paix l' année suivante
Wikipedia.fr-en.chunks.fr:au moins un juge de paix
Wikipedia.fr-en.chunks.fr:juge de paix et leader d' une unité de milice locale pendant
Wikipedia.fr-en.chunks.fr:, juge de paix
Wikipedia.fr-en.chunks.fr:juge de paix au
Wikipedia.fr-en.chunks.fr:chaque juge de paix du Lancashire
Wikipedia.fr-en.chunks.fr:l' intervention du juge de paix
Wikipedia.fr-en.chunks.fr:il est nommé juge de paix
Wikipedia.fr-en.chunks.fr:premier marchand général et premier juge de paix

```

**Comment by philippehenri** [comment]
hmm i see. ps: i note that you have quite a few duplicates there also... so maybe it helps :)

**Comment by EtienneAb3d** [comment]
The duplications are those of the original sentences. It thus reflect the statistical weight of each in the corpus.
